io_scene_godot/converters/mesh.py

```diff
@@ -67,6 +69,10 @@ def export_mesh_node(escn_file, export_settings, obj, parent_gd_node):
             escn_file, export_settings, mesh_node,
             obj.data.shape_keys, 'shapekey')
 
+    if ("ARMATURE" in export_settings['object_types'] and
```

Comment:
I think it would be better to merge this block to the above `if ("ARMATURE"` block, which would seem to have more continuous logic. 

It can be merged writing as:
``` python
if AMRATURE_CHECK:
     # set to REST
     ...
     mesh_id = mesh_exporter.export_mesh(escn_file, export_settings)
     # set back
     ...
else:
     mesh_id = mesh_exporter.export_mesh(escn_file, export_settings)
```


Still, the current implementation is good enough, it depends on you to change it or not, I would merge it anyway :)

Reply:
Good points, will implement the specified changes.

Just curious if there's a good reason why we need to evaluate the object, to_mesh() it and triangulate it twice if it's a physics objects?
Why not do all that stuff at the top of export_mesh_node() ?

Reply:
Also, this:

`# Check if it exists so we don't bother to export it twice
        mesh = self.object.data
        mesh_id = escn_file.get_internal_resource(mesh)`

Doesn't take into account the mesh's modifiers. The same mesh data block can be used on multiple objects, but the modifiers could produce different meshes. This check would ignore the mesh objects with different modifers, if the mesh data block has already been processed.

Another reason to evaluate the object beforehand I think.

Reply:
> Just curious if there's a good reason why we need to evaluate the object, to_mesh() it and triangulate it twice if it's a physics object?

Good question, it might be twice, but ideally it would be just once. First, in Blender there is no one to one mapping of Godot's CollisionShape node. In order to support it, the exporter export any Mesh node 
1. with `Rigid Body` 
2. draw type is `WIRE` 

as a physics node subtree(`Rigid` + `CollisionShape`).  Any Mesh satisfies both conditions break from the mesh node exporting [here](https://github.com/godotengine/godot-blender-exporter/blob/blender2.8/io_scene_godot/converters/mesh.py#L29). This is the idea way of creating a CollisionShape in Blender and mesh get exported once.

> Why not do all that stuff at the top of export_mesh_node(), send the resulting eval_obj to physics.export_physics_properties() and MeshResourceExporter()?

Good point, maybe we could make a abstract class namely `MeshNodeExporter` and inherits to a `PhysicalMeshExporter` and `MeshResourceExporter` so that some code could be shared.

Reply:
>The same mesh data block can be used on multiple objects, but the modifiers could produce different meshes. This check would ignore the mesh objects with different modifers, if the mesh data block has already been processed.
> 

Good catch! This is definitely a bug, I will create an issue later and think about it more.

Reply:
I've tried to address some of the problems we've been discussing here, but it's not as neat as I'd hoped.
I've added a helper class - MeshConverter - which converts objects to meshes, applying modifiers (when specified), also handles triangulation and calculating tangents (when specified), reducing a bit of repetition across mesh.py and physics.py.
I've also re-arranged a few things and made some common function arguments in like escn_file and export_settings class variables in MeshResourceExporter because they're referred to so much.

However I'm holding off from pushing any changes until Blender fixes its API, there are bugs in all the recent blender builds that are making a fully functional exporter impossible right now.

Yesterday I realised evaluated_get on objects with modifiers was resulting in objects with empty vertex groups. This was fixed last night and the API was updated, however these new changes have resulted in a new bug, whereby Object.to_mesh() produces a mesh with no materials.

It's all a bit screwy right now, I wouldn't be surprised if they implement another overhaul on the whole depsgraph / to_mesh mess.

Reply:
Btw is there a problem if 2 resources in an escn have the same name?
As I mentioned before, with modifiers enabled it's possible you may use the same blender mesh more than once and create 2 unrelated meshes in godot, but the exporter uses mesh.name for mesh-related resource names which could result in names clashing. I've implemented a way to identify potential problem meshes and append them with some random alphanumeric string to prevent clashes, just wondering though if that's necessary?

Reply:
> However I'm holding off from pushing any changes until Blender fixes its API, there are bugs in all the recent blender builds that are making a fully functional exporter impossible right now.

Yeah, just leave them until the APIs got stable. Thank you for your work and following the blender devs.

> I've added a helper class - MeshConverter - which converts objects to meshes, applying modifiers (when specified), also handles triangulation and calculating tangents (when specified), reducing a bit of repetition across mesh.py and physics.py.
I've also re-arranged a few things and made some common function arguments in like escn_file and export_settings class variables in MeshResourceExporter because they're referred to so much.

Wow, really appreciate it you doing this, it could be beneficial. I suggest you doing this refactoring in another PR, which would make the review and test easier

> Btw is there a problem if 2 resources in an escn have the same name?
No, actually the resource are directly using a blender object (in example it's `mesh`) as hash key, like
``` python
        mesh = self.object.data
        mesh_id = escn_file.get_internal_resource(mesh)
```
blender object is actually a C pointer, so name does not affect.

Reply:
> No, actually the resource are directly using a blender object (in example it's `mesh`) as hash key, like
> 
> ```python
>         mesh = self.object.data
>         mesh_id = escn_file.get_internal_resource(mesh)
> ```
> 
> blender object is actually a C pointer, so name does not affect.

I was referring to the resource_name property in a sub_resource.
I was wondering if this was ok:

```
[sub_resource id=1 type="ArrayMesh"]
resource_name = "Cube"
...
[sub_resource id=2 type="ArrayMesh"]
resource_name = "Cube"
...
```

I'm not sure, so I'm appending potential problem mesh resources with random alphanumeric strings just in case.

> I suggest you doing this refactoring in another PR, which would make the review and test easier

OK, I'll revert those particular changes for now as they're not significant, but there's a fair amount of difference between the last commit, but I've tried to keep mesh.py as similar as possible to the original.

Also, the Blender API bug where materials weren't carrying over after calling to_mesh() has just been fixed, so it should be released in tonight's beta release.

> Wow, really appreciate it you doing this, it could be beneficial.

And I really appreciate the amount of effort that you've put into this project. I haven't come across another blender exporter addon that manages to convert a 3D scene into another format so comprehensively. The material exporter in particular is excellent.
It's an invaluable tool for those of us working in Blender.

Reply:
> I was referring to the resource_name property in a sub_resource.

Ah sorry for my misunderstanding. Yeah, you have a point. However, I did some test on, both Blender and Godot can resolve the resource name conflict, so we don't need to do anything about it.

> OK, I'll revert those particular changes for now as they're not significant, but although I've tried to keep mesh.py as similar as possible to the original, there's a fair amount of difference between my current version and the last commit.

You may not have to revert, you can use `git cherry-pick` to pick commits from one branch to another.

Thank you for your contribution as well ^ ^

Reply:
Correction:

Yesterday I mentioned 
>Godot can resolve the resource name conflict

which may not be true, if the mesh import option set to `Files`, mesh resources with same name may overwrites each other and cause the error. However, blender seems has a strict checking against Mesh with same name, so this `resource_name` is still not important IMO